tinygraph/tinygraph-index.c

```diff
@@ -0,0 +1,538 @@
+}
+
+
+void tinygraph_index_destruct(tinygraph_index * const index) {
+  if (!index) {
+    return;
+  }
+
+  free(index->lats);
+  free(index->lngs);
+  free(index->zvals);
+  free(index->nodes);
+
+  free(index);
+}
+
+
+static inline const uint64_t* tinygraph_index_bsearch_lt(
+    const uint64_t* first,
+    const uint64_t* last,
+    uint64_t value
+) {
+  TINYGRAPH_ASSERT(first < last);
+
+  const uint64_t *it = first;
+
+  uint64_t step = 0, n = last - first;
+
+  while (n > 0) {
+    step = n / 2;
+    it = first + step;
+
+    if (*it < value) { // <- lt means less than here
+      first = it + 1;
+      n -= step + 1;
+    } else {
+      n = step;
+    }
+  }
+
+  return first;
+}
+
+static inline const uint64_t* tinygraph_index_bsearch_lte(
+    const uint64_t* first,
+    const uint64_t* last,
+    uint64_t value
+) {
+  TINYGRAPH_ASSERT(first < last);
+
+  const uint64_t *it = first;
+
+  uint64_t step = 0, n = last - first;
+
+  while (n > 0) {
+    step = n / 2;
+    it = first + step;
+
+    if (*it <= value) { // <- lte means less than or equal here
+      first = it + 1;
+      n -= step + 1;
+    } else {
+      n = step;
+    }
+  }
+
+  return first;
+}
+
+// With a z-value that it outside a bounding box, find the
+// smallest next z-value that is again within a bounding box.
+// This function goes back to the 80s and has various names:
+// BIGMIN/LITMAX, nextJumpIn, GetNextZ-Address
+//
+// The following implementation is based on the original
+// Tropf and Herzog paper linked above and by taking
+// inspiration from https://github.com/smatsumt/pyzorder
+// with MIT License, Copyright (c) 2019 Shinji Matsumoto.
+static inline uint64_t tinygraph_index_bigmin(uint64_t zval, uint64_t zmin, uint64_t zmax) {
+  TINYGRAPH_ASSERT(zval >= zmin);
+  TINYGRAPH_ASSERT(zval < zmax);
+  TINYGRAPH_ASSERT(zmin <= zmax);
+
+  uint64_t bigmin = zmin;
+
+  uint64_t load_mask = UINT64_C(0x5555555555555555);
+  uint64_t load_ones = UINT64_C(0x2aaaaaaaaaaaaaaa);
+
+  uint64_t mask = UINT64_C(0x8000000000000000);
+
+  while (mask) {
+    const uint64_t bzval = zval & mask;
+    const uint64_t bzmin = zmin & mask;
+    const uint64_t bzmax = zmax & mask;
+
+    if (!bzval && !bzmin && !bzmax) {
+      // pass
+    } else if (!bzval && !bzmin && bzmax) {
+      bigmin = (zmin & load_mask) | mask;
+      zmax = (zmax & load_mask) | load_ones;
+    } else if (!bzval && bzmin && bzmax) {
+      TINYGRAPH_ASSERT(zmin > zval);
+      return zmin;
+    } else if (bzval && !bzmin && !bzmax) {
+      TINYGRAPH_ASSERT(bigmin > zval);
+      return bigmin;
+    } else if (bzval && !bzmin && bzmax) {
+      zmin = (zmin & load_mask) | mask;
+    } else if (bzval && bzmin && bzmax) {
+      // pass
+    } else {
+      TINYGRAPH_ASSERT(false);
+      TINYGRAPH_UNREACHABLE();
+    }
+
+    mask >>= 1;
+    load_ones >>= 1;
+    load_mask >>= 1;
+    load_mask |= UINT64_C(0x8000000000000000);
+  }
+
+  TINYGRAPH_ASSERT(bigmin > zval);
+  return bigmin;
+}
```

Comment:
This is the BIGMIN/jumpNextIn/GetNextZ-Address optimization.

With a z value on the z-order space filling curve outside of a query bounding box, it computes the next valid z value inside the bounding box. This allows us to prune the search space and scan sub-ranges in a sorted array of z-values.